File: 1_Array_And_Hashing/group_anagrams.py
import collections

class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        # Initialize a defaultdict to store anagrams grouped by their character counts
        answer = collections.defaultdict(list)

        # Iterate through each string in the input list
        for char in strs:
            # Initialize a list to count the occurrences of each character in the string
            count = [0] * 26

            # Count the occurrences of each character in the string
            for letter in char:
                # using ASCII to map lowercase letter to a number between 97 - 122 (a - z)
                count[ord(letter) - ord("a")] += 1

            # Convert the count list to a tuple to use it as a key in the defaultdict
            answer[tuple(count)].append(char)

        # Return the grouped anagrams as a list of lists
        return answer.values()
    

# implementation
print("Leetcode 49: Group Anagrams\n")
answer = Solution()

# Example 1
example_1 = ["eat", "tea", "tan", "ate", "nat", "bat"]
print(f"Example 1: strs = {example_1}")
print("Outcome:", answer.groupAnagrams(example_1))

# Example 1
example_2 = [""]
print(f"Example 2: strs = {example_2}")
print("Outcome:", answer.groupAnagrams(example_2))

# Example 3
example_3 = ["a"]
print(f"Example 3: strs = {example_3}")
print("Outcome:", answer.groupAnagrams(example_3))

File: 1_Array_And_Hashing/group_anagrams.go
package main

import (
	"fmt"
	"sort"
)

func groupAnagrams(strs []string) [][]string {
	groupedAnagrams := make(map[string][]string)

	for _, str := range strs {
		// Convert string to slice of bytes
		slice := []byte(str)
		// Sort the slice of bytes
		sort.Slice(slice, func(i, j int) bool {
			return slice[i] < slice[j]
		})
		// Convert the sorted slice back to string
		sortedStr := string(slice)

		// If the sorted string is not in the map, initialize a new slice
		if _, ok := groupedAnagrams[sortedStr]; !ok {
			groupedAnagrams[sortedStr] = []string{}
		}
		// Append the original string to the slice in the map
		groupedAnagrams[sortedStr] = append(groupedAnagrams[sortedStr], str)
	}

	// Convert map values to slice of slices
	result := make([][]string, 0, len(groupedAnagrams))
	for _, v := range groupedAnagrams {
		result = append(result, v)
	}

	return result
}

func main() {
	fmt.Println("Group Anagrams\n")

	// Example 1
	example1 := []string{"eat", "tea", "tan", "ate", "nat", "bat"}
	fmt.Println("Example 1: strs = ['eat', 'tea', 'tan', 'ate', 'nat', 'bat']")
	fmt.Println("Output:", groupAnagrams(example1))

	// Example 2
	example2 := []string{""}
	fmt.Println("Example 2: strs = ['']")
	fmt.Println("Output:", groupAnagrams(example2))

	// Example 3
	example3 := []string{"a"}
	fmt.Println("Example 2: strs = ['a']")
	fmt.Println("Output:", groupAnagrams(example3))
}
